software/apps/cfft_radix4_q16/main.c

```diff
@@ -0,0 +1,214 @@
+// Copyright 2022 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Author: Marco Bertuletti, ETH Zurich
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* Mempool runtime libraries */
+#include "encoding.h"
+#include "printf.h"
+#include "runtime.h"
+#include "synchronization.h"
+#include "xpulp/builtins_v2.h"
+
+/* CFFT data libraries */
+#include "data/data_cfft_radix4_q16.h"
+
+/*
+  CHOOSE ONE
+   - SINGLE:    Single core FFT
+   - PARALLEL:  Parallel FFT not "memory-aware"
+   - FOLDED:    Parallel FFT with "memory-aware" load/store scheme
+   - SCHEDULED: Scheduling of multiple parallel FFTs with "memory-aware"
+  load/store scheme
+      - N_FFTs_COL: Independent FFTs scheduled on one row (default 1)
+      - N_FFTs_ROW: Independent FFTs scheduled on columns (default 1)
+      (OPTIONALLY ENABLE)
+      - FOLDED_TWIDDLES: Also the twiddles have "memory-aware" load/stores
+      - BITREVERSETABLE: The bitreversal indeces are loaded from a table
+      - ASM:             Use asm_volatile statements
+*/
+
+#define FOLDED
+#define FOLDED_TWIDDLES
+#define BITREVERSETABLE
+#define ASM // Use asm_volatile statements
+
+#if !(defined(N_FFT_ROW) && defined(N_FFTs_COL))
+#define N_FFTs_ROW 1
+#define N_FFTs_COL 1
+#endif
+
+#define ABS(x) (((x) < 0) ? (-x) : (x))
+#include "kernel/mempool_radix4_cfft_butterfly_q16.h"
+#include "kernel/mempool_radix4_cfft_q16_bitreversal.h"
+#include "kernel/mempool_radix4_cfft_q16p.h"
+#include "kernel/mempool_radix4_cfft_q16s.h"
+
+int16_t pSrc[N_FFTs_ROW * 8 * N_BANKS]
+    __attribute__((aligned(N_FFTs_ROW * 8 * N_BANKS), section(".l1")));
+int16_t pDst[N_FFTs_ROW * 8 * N_BANKS]
+    __attribute__((aligned(N_FFTs_ROW * 8 * N_BANKS), section(".l1")));
+int16_t pCoef16_src[8 * N_BANKS]
+    __attribute__((aligned(8 * N_BANKS), section(".l1")));
+int16_t pCoef16_dst[8 * N_BANKS]
+    __attribute__((aligned(8 * N_BANKS), section(".l1")));
+uint16_t pRevT16[BITREVINDEXTABLE_FIXED_TABLE_LENGTH]
+    __attribute__((aligned(N_BANKS), section(".l1")));
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+/* INITIALIZATION FUNCTIONS*/
+
+void initialize_l1() {
+  uint32_t core_id = mempool_get_core_id();
+  uint32_t num_cores = mempool_get_core_count();
+
+  // Initialize the inputs and results from the data.h file
+  for (uint32_t j = 0; j < N_FFTs_ROW; j++) {
+    for (uint32_t i = core_id; i < (8 * N_BANKS); i += num_cores) {
+      if (i < N_RSAMPLES * N_FFTs_COL) {
+        pSrc[j * (8 * N_BANKS) + i] = (int16_t)vector_inp[i % N_RSAMPLES];
+      } else {
+        pSrc[j * (8 * N_BANKS) + i] = (int16_t)0;
+      }
+      pDst[j * (8 * N_BANKS) + i] = (int16_t)0;
+    }
+  }
```

Comment:
What is the reasoning behind initializing `N_FFT_ROWS * 8 * N_BANKS`? Where are the 8 and banks coming from, and why do we initialize everything beyond the `NRSAMPLES` with zero and the output also with zero?

Reply:
When we fold the FFT each radix-4 wing of a butterfly will take a different row of the memory, so four rows in total. Since we have complex data we skip 2 * 4 * N_BANKS. The data is not copied folded though, as we will still copy the data from different columns one after the other and then fold it in the kernel.